### PR TITLE
Refactor subsolver setup

### PR DESCRIPTION
**Pull request by ajreynol**
This refactors the interfaces for spawning subsolvers so that necessary information, e.g. the separation logic heap, is considered.

This makes SyGuS + separation logic possible, a regression is added.

### CODE REVIEW

File: src/theory/smt_engine_subsolver.cpp [hunk 1]
```diff
@@ -42,40 +69,37 @@ Result quickCheck(Node& query)
...
-  initializeSubsolver(
-      smte, env.getOptions(), env.getLogicInfo(), needsTimeout, timeout);
-  // set up separation logic heap if necessary
-  TypeNode sepLocType, sepDataType;
-  if (env.getSepHeapTypes(sepLocType, sepDataType))
-  {
-    smte->declareSepHeap(sepLocType, sepDataType);
-  }
+  SubsolverSetupInfo ssi(env);
+  initializeSubsolver(smte, ssi, needsTimeout, timeout);
```
Comment (4tXJ7f) [inline src/theory/smt_engine_subsolver.cpp:97]: This overload should not be necessary. `initializeSubsolver(..., d_env, ...)` should work as expected due to implicit construction of the `SubsolverSetupInfo` instance.

File: src/theory/quantifiers/sygus/sygus_repair_const.cpp [hunk 2]
```diff
@@ -230,10 +230,10 @@ bool SygusRepairConst::repairSolution(Node sygusBody,
   Trace("sygus-engine") << "Repairing previous solution..." << std::endl;
   // make the satisfiability query
   std::unique_ptr<SolverEngine> repcChecker;
+  SubsolverSetupInfo ssi(d_env);
   // initialize the subsolver using the standard method
   initializeSubsolver(repcChecker,
-                      d_env.getOptions(),
-                      d_env.getLogicInfo(),
+                      ssi,
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/sygus_repair_const.cpp:236]:
```suggestion
                      d_env,
```

File: src/theory/quantifiers/sygus/sygus_repair_const.cpp [hunk 3]
```diff
@@ -230,10 +230,10 @@ bool SygusRepairConst::repairSolution(Node sygusBody,
   Trace("sygus-engine") << "Repairing previous solution..." << std::endl;
   // make the satisfiability query
   std::unique_ptr<SolverEngine> repcChecker;
+  SubsolverSetupInfo ssi(d_env);
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/sygus_repair_const.cpp:233]:
```suggestion
```

File: src/theory/quantifiers/sygus/cegis_core_connective.cpp [hunk 4]
```diff
@@ -564,7 +564,8 @@ Result CegisCoreConnective::checkSat(Node n, std::vector<Node>& mvs) const
 {
   Trace("sygus-ccore-debug") << "...check-sat " << n << "..." << std::endl;
   n = rewrite(n);
-  Result r = checkWithSubsolver(n, d_vars, mvs, options(), logicInfo());
+  SubsolverSetupInfo ssi(d_env);
+  Result r = checkWithSubsolver(n, d_vars, mvs, ssi);
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/cegis_core_connective.cpp:568]:
```suggestion
  Result r = checkWithSubsolver(n, d_vars, mvs, d_env);
```

File: src/theory/quantifiers/expr_miner.cpp [hunk 5]
```diff
@@ -57,26 +57,23 @@ Node ExprMiner::convertToSkolem(Node n)
 void ExprMiner::initializeChecker(std::unique_ptr<SolverEngine>& checker,
                                   Node query)
 {
-  initializeChecker(checker, query, options(), logicInfo());
+  SubsolverSetupInfo ssi(d_env);
+  initializeChecker(checker, query, ssi);
```
Comment (4tXJ7f) [inline src/theory/quantifiers/expr_miner.cpp:61]:
```suggestion
  initializeChecker(checker, query, d_env);
```

**Review by 4tXJ7f — APPROVED** [review]
LGTM, just some minor suggestions.

File: src/theory/smt_engine_subsolver.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/smt_engine_subsolver.cpp:97]: I'd prefer to have this be written explicitly.  I don't like the idea of having explicit construction happening in this case.

**Pull request merged**